jaxley/connect.py

```diff
@@ -80,45 +99,50 @@ def sparse_connect(
     post_cell_view: "View",
     synapse_type: "Synapse",
     p: float,
+    random_post_comp: bool = False,
 ):
     """Appends multiple connections which build a sparse, randomly connected layer.
 
-    Connections are from branch 0 location 0 to a randomly chosen branch and loc.
+    Connections are from branch 0 location 0 of the pre-synaptic cell to branch 0
+    location 0 of the post-synaptic cell unless random_post_comp=True.
 
     Args:
         pre_cell_view: View of the presynaptic cell.
         post_cell_view: View of the postsynaptic cell.
         synapse_type: The synapse to append.
         p: Probability of connection.
+        random_post_comp: If True, randomly samples the postsynaptic compartments.
     """
     # Get pre- and postsynaptic cell indices.
-    pre_cell_inds = pre_cell_view._cells_in_view
-    post_cell_inds = post_cell_view._cells_in_view
-    num_pre = len(pre_cell_inds)
-    num_post = len(post_cell_inds)
-
-    num_connections = np.random.binomial(num_pre * num_post, p)
-    pre_syn_neurons = np.random.choice(pre_cell_inds, size=num_connections)
-    post_syn_neurons = np.random.choice(post_cell_inds, size=num_connections)
-
-    # Sort the synapses only for convenience of inspecting `.edges`.
-    sorting = np.argsort(pre_syn_neurons)
-    pre_syn_neurons = pre_syn_neurons[sorting]
-    post_syn_neurons = post_syn_neurons[sorting]
-
-    # Post-synapse is a randomly chosen branch and compartment.
-    global_post_indices = [
-        sample_comp(post_cell_view.scope("global").cell(cell_idx))
-        for cell_idx in post_syn_neurons
-    ]
-    global_post_indices = (
-        np.hstack(global_post_indices) if len(global_post_indices) > 1 else []
-    )
-    post_rows = post_cell_view.base.nodes.loc[global_post_indices]
-
-    # Pre-synapse is at the zero-eth branch and zero-eth compartment.
-    global_pre_indices = pre_cell_view.base._cumsum_nseg_per_cell[pre_syn_neurons]
-    pre_rows = pre_cell_view.base.nodes.loc[global_pre_indices]
+    num_pre = len(pre_cell_view._cells_in_view)
+    num_post = len(post_cell_view._cells_in_view)
+
+    # Generate random cxns without duplicates --> respects p but memory intesive if extremely large n cells
```

Comment:
When I implemented the connect API, @michaeldeistler and me had discussed not wanting to do this, since this can be really bad for large, sparse networks. Did we change our mind about doing it this way?

Reply:
I agree its cleaner and easier to parse, but not efficient.

Reply:
Good catch Jonas! Let's avoid building a `n x n` matrix.

Reply:
I was looking into this for quite a while, and all of the ways that are memory efficient are extremely time intensive, and vice versa, so I came up with a way to split the cost a bit... generating blocks of random matrix at a time. I am still thinking of how it could be prettier, but the resource management seems to be good

Reply:
The function before created many duplicate connections between the same cells, and simply filtering these out would remove a lot of connections (inefficient) and make the p value entered somewhat meaningless (more problematic). Also p=1.0 did not create a fully connected graph, far from it I might add. Resource efficient methods for constructing sparse random graphs are highly sought after and typically trade off computation speed and memory, so I tried to write a function that balances the two. The function now isn't O(N^2) per se, the largest matrix it creates is 100x100. I can keep looking for better algorithms in the future, but for now the function at least fixes the bug of duplicate connections and restores the traditional meaning of p.

Reply:
I'm also thinking now that just removing sparse_connect() might not be so bad... I think we might have used connectivity matrix connect or something else in the RNN experiments anyway